pkg/sanity/controller.go

```diff
@@ -1580,18 +1516,135 @@ var _ = DescribeSanity("ExpandVolume [Controller Server]", func(sc *TestContext)
 		vol := r.MustCreateVolume(context.Background(), req)
 
 		By("expanding the volume")
-		expReq := &csi.ControllerExpandVolumeRequest{
+		expReq := MakeExpandVolumeReq(sc, vol.GetVolume().GetVolumeId())
+		rsp, err := r.ControllerExpandVolume(context.Background(), expReq)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(rsp).NotTo(BeNil())
+		Expect(rsp.GetCapacityBytes()).To(Equal(TestVolumeExpandSize(sc)))
+	})
+})
+
+var _ = DescribeSanity("ModifyVolume [Controller Server]", func(sc *TestContext) {
+	var r *Resources
+
+	BeforeEach(func() {
+		r = &Resources{
+			ControllerClient: csi.NewControllerClient(sc.ControllerConn),
+			Context:          sc,
+		}
+
+		if !isControllerCapabilitySupported(r, csi.ControllerServiceCapability_RPC_MODIFY_VOLUME) {
+			Skip("ControllerExpandVolume not supported")
+		}
+	})
+
+	AfterEach(func() {
+		r.Cleanup()
+	})
+
+	It("should fail if no volume id is given", func() {
+		modifyReq := MakeModifyVolumeReq(sc, "")
+		rsp, err := r.ControllerModifyVolume(context.Background(), modifyReq)
+		ExpectErrorCode(rsp, err, codes.InvalidArgument)
+	})
+
+	It("should fail if volume does not exist", func() {
+		modifyReq := MakeModifyVolumeReq(sc, "non-existing-volume-id")
+		rsp, err := r.ControllerModifyVolume(context.Background(), modifyReq)
+		ExpectErrorCode(rsp, err, codes.NotFound)
+	})
+
+	It("should fail if specified mutable parameters are not supported by the volume", func() {
+
+		By("creating a new volume")
+
+		volReq := MakeCreateVolumeReq(sc, UniqueString("sanity-modify-volume"))
+		vol := r.MustCreateVolume(context.Background(), volReq)
+
+		By("failing to modify the volume")
+
+		modifyReq := &csi.ControllerModifyVolumeRequest{
 			VolumeId: vol.GetVolume().GetVolumeId(),
-			CapacityRange: &csi.CapacityRange{
-				RequiredBytes: TestVolumeExpandSize(sc),
+			MutableParameters: map[string]string{
+				"XXX_FakeKey": "XXX_FakeValue",
 			},
-			Secrets:          sc.Secrets.ControllerExpandVolumeSecret,
-			VolumeCapability: TestVolumeCapabilityWithAccessType(sc, csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER),
+			Secrets: sc.Secrets.ControllerModifyVolumeSecret,
 		}
-		rsp, err := r.ControllerExpandVolume(context.Background(), expReq)
+		rsp, err := r.ControllerModifyVolume(context.Background(), modifyReq)
+		ExpectErrorCode(rsp, err, codes.InvalidArgument)
+	})
+
+	// TODO Q: Should we enforce Empty MutableParameters list as InvalidArgument Err?
+	// TODO Q Cont: KEP says it is OPTIONAL, k/k: storage/types says required, but no mention in CSI Spec
+	It("should fail if no mutable parameters specified", func() {
```

Comment:
CSI spec says mutable_parameters is optional so empty MutableParameters list is fine.
```
  // Plugin specific creation-time parameters passed in as opaque 
  // key-value pairs. These mutable_parameteres MAY also be
  // changed during the lifetime of the volume via a subsequent
  // `ControllerModifyVolume` RPC. This field is OPTIONAL.
  // The Plugin is responsible for parsing and validating these
  // parameters. COs will treat these as opaque.
  
  // Plugins MUST treat these 
  // as if they take precedence over the parameters field.
  // This field SHALL NOT be specified unless the SP has the
  // MODIFY_VOLUME plugin capability.
  map<string, string> mutable_parameters = 8 [(alpha_field) = true];
```